src/main/java/val/ResultMap.java

```diff
@@ -0,0 +1,167 @@
+	if (results == null)
+	    throw new NullPointerException("results must not be null");
+	Map<String, List<Result>> map = new HashMap<>();
+	map.put(key, results);
+	return ResultMap.from(map);
+    }
+
+    /**
+     * Internal constructor which ensures immutability of contained content.
+     */
+    private ResultMap(@NonNull Map<String, List<Result>> arg) {
+	// Validate that provided map does not contain any null pointers,
+	// while copying to a structure that can be typed as such.
+	Map<String, List<Result>> values =
+	    new HashMap<String, List<Result>>(arg.size());
+	for (Map.Entry<String, List<Result>> entry: arg.entrySet()) {
+	    if (entry.getKey() == null)
+		throw new NullPointerException("Null keys are prohibited.");
+	    List<Result> from = entry.getValue();
+	    if (from == null)
+		throw new NullPointerException("Null values are prohibited.");
+	    List<Result> results = new ArrayList<Result>(from.size());
+	    for (Result result: from) {
+		if (result == null)
+		    throw new NullPointerException("Null values are prohibited");
+		results.add(result);
+	    }
+	    values.put(entry.getKey(), Collections.unmodifiableList(results));
+	}
+	this.values = Collections.unmodifiableMap(values);
+	clean = this.values.isEmpty();
+    }
+
+    /**
+     * Return a representation of this ResultMap as a Map.
+     *
+     * @return A Map containing the entries within this ReulstMap.
+     */
+    public Map<String, List<Result>> asMap() {
+	return values;
+    }
+
+    /**
+     * Get the combined sum of the current ResultMap with the provided argument.
+     *
+     * Return a ResultMap which contains the merged result of the entires of
+     * this ResultMap and the one passed as an argument.
+     *
+     * @param right The ResultMap to merge with this ResultMap
+     * @return A ResultMap containing the combined results of this and right.
+     * @throws NullPointerException if right is null.
+     */
+    @EnsuresNonNull({"#1"})
+    ResultMap plus(ResultMap right) {
+	if (right == null)
+	    throw new NullPointerException("Null argument is prohibited.");
+	if (right.isClean()) return this;
+	if (this.isClean()) return right;
+	Map<String, List<Result>> merged = new HashMap<>(this.values.size());
+	mapCopy(merged, this.values);
+	mapCopy(merged, right.values);
+	return new ResultMap(merged);
+    }
+    // Facilitate deep copies
+    private void mapCopy(Map<String, List<Result>> dest,
+			 Map<String, List<Result>> src) {
+	for (Map.Entry<String, List<Result>> e: src.entrySet()) {
+	    String k = e.getKey();
+	    List<Result> vals
+		= dest.getOrDefault(k, new ArrayList<>(e.getValue().size()));
+	    vals.addAll(e.getValue());
+	    dest.put(k, vals);
+	}
+    }
+
+    @Override
+    public boolean equals(@Nullable Object other) {
+	return other instanceof ResultMap &&
+	    Objects.equals(values, ((ResultMap) other).values);
+    }
+
+    @Override
+    public int hashCode() {
+	return values.hashCode();
```

Comment:
nit: but is this safe? doesn't that mean a ResultCode and its internal values map would both have the same hashCode? i suppose that doesn't really cause a problem as long as equals is false in that case (which it would be)

Reply:
I think the difference in types should be sufficient? (I can't think of a reason it wouldn't be but I could definitely be missing something).

Reply:
This should satisfy the contract (so the paranoia you gave me is gone). https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#hashCode()